File: Typescript/mcp-server/alert-mcp-server.ts
/**
 * Alert MCP Server — TypeScript/Node.js
 *
 * Exposes 3 tools to GitHub Copilot via the Model Context Protocol (MCP):
 *   get_alerts(severity?)   → query the in-memory alert store
 *   get_alert_stats()       → severity breakdown count
 *   create_alert(...)       → add a new alert
 *
 * Transport: Stdio (default — VS Code spawns this as a child process)
 * To switch to HTTP Streamable or SSE, run /MCP-Server-Builder in Copilot Chat.
 *
 * Install & run:
 *   npm install
 *   npx ts-node alert-mcp-server.ts
 */

import { Server } from "@modelcontextprotocol/sdk/server/index.js";
import { StdioServerTransport } from "@modelcontextprotocol/sdk/server/stdio.js";
import {
  CallToolRequestSchema,
  ListToolsRequestSchema,
} from "@modelcontextprotocol/sdk/types.js";

// ── Types ────────────────────────────────────────────────────────────────────

type Severity = "critical" | "high" | "medium" | "low";
type AlertStatus = "open" | "acknowledged" | "resolved";

interface Alert {
  id: string;
  title: string;
  severity: Severity;
  status: AlertStatus;
  source: string;
  timestamp: string;
}

// ── In-memory alert store (POC) ───────────────────────────────────────────────
// TODO: Use the /generate-alerts skill to populate with more realistic data

const alertStore: Alert[] = [
  { id: "1", title: "High CPU usage on prod-01", severity: "critical", status: "open", source: "monitoring", timestamp: new Date().toISOString() },
  { id: "2", title: "Disk space below 10% on db-server", severity: "high", status: "open", source: "infrastructure", timestamp: new Date().toISOString() },
  { id: "3", title: "API response latency spike (p99 > 2s)", severity: "medium", status: "acknowledged", source: "api-gateway", timestamp: new Date().toISOString() },
  { id: "4", title: "SSL certificate expiring in 7 days", severity: "low", status: "open", source: "security", timestamp: new Date().toISOString() },
];

// ── MCP Server ────────────────────────────────────────────────────────────────

const server = new Server(
  { name: "alert-mcp-server", version: "1.0.0" },
  { capabilities: { tools: {} } }
);

// ── Tool Definitions ──────────────────────────────────────────────────────────

server.setRequestHandler(ListToolsRequestSchema, async () => ({
  tools: [
    {
      name: "get_alerts",
      description: "Get alerts from the alert management system, optionally filtered by severity",
      inputSchema: {
        type: "object",
        properties: {
          severity: {
            type: "string",
            enum: ["critical", "high", "medium", "low"],
            description: "Filter by severity (optional — omit to get all alerts)",
          },
        },
      },
    },
    {
      name: "get_alert_stats",
      description: "Get a summary count of alerts broken down by severity level",
      inputSchema: { type: "object", properties: {} },
    },
    {
      name: "create_alert",
      description: "Create a new alert in the alert management system",
      inputSchema: {
        type: "object",
        required: ["title", "severity"],
        properties: {
          title: { type: "string", description: "Short description of the alert" },
          severity: {
            type: "string",
            enum: ["critical", "high", "medium", "low"],
          },
          source: {
            type: "string",
            description: "System or service that triggered the alert (optional)",
          },
        },
      },
    },
  ],
}));

// ── Tool Handlers ─────────────────────────────────────────────────────────────

server.setRequestHandler(CallToolRequestSchema, async (request) => {
  const { name, arguments: args } = request.params;

  switch (name) {
    case "get_alerts": {
      const results = args?.severity
        ? alertStore.filter((a) => a.severity === args.severity)
        : alertStore;
      return {
        content: [{ type: "text", text: JSON.stringify(results, null, 2) }],
      };
    }

    case "get_alert_stats": {
      const bySeverity = alertStore.reduce((acc, a) => {
        acc[a.severity] = (acc[a.severity] ?? 0) + 1;
        return acc;
      }, {} as Record<string, number>);
      return {
        content: [
          { type: "text", text: JSON.stringify({ total: alertStore.length, bySeverity }, null, 2) },
        ],
      };
    }

    case "create_alert": {
      // TODO: Ask Copilot to add input validation using inline chat (/fix)
      const newAlert: Alert = {
        id: String(alertStore.length + 1),
        title: args!.title as string,
        severity: args!.severity as Severity,
        status: "open",
        source: (args?.source as string) ?? "manual",
        timestamp: new Date().toISOString(),
      };
      alertStore.push(newAlert);
      return {
        content: [{ type: "text", text: `✅ Alert created:\n${JSON.stringify(newAlert, null, 2)}` }],
      };
    }

    default:
      throw new Error(`Unknown tool: ${name}`);
  }
});

// ── TRANSPORT — swap this block to switch transport modes ─────────────────────
// See /MCP-Server-Builder prompt for HTTP Streamable or SSE alternatives

async function main() {
  const transport = new StdioServerTransport();
  await server.connect(transport);
  console.error("🚀 Alert MCP Server running (stdio transport)");
}

main().catch(console.error);

File: dotNet/McpServer/AlertMcpServer.cs
/*
 * Alert MCP Server — C#/.NET
 *
 * Exposes 3 tools to GitHub Copilot via the Model Context Protocol (MCP):
 *   get_alerts(severity?)   → query the in-memory alert store
 *   get_alert_stats()       → severity breakdown count
 *   create_alert(...)       → add a new alert
 *
 * Transport: Stdio (default — VS Code spawns this as a child process)
 * To switch to HTTP Streamable, run /MCP-Server-Builder in Copilot Chat.
 *
 * Install & run:
 *   dotnet restore
 *   dotnet run
 */

using ModelContextProtocol.Server;
using System.ComponentModel;
using System.Text.Json;

// ── In-memory alert store (POC) ───────────────────────────────────────────────
// TODO: Use the /generate-alerts skill to populate with more realistic data

var alertStore = new List<Alert>
{
    new("1", "High CPU usage on prod-01",       "critical", "open",         "monitoring",     DateTime.UtcNow),
    new("2", "Disk space below 10% on db-server","high",    "open",         "infrastructure", DateTime.UtcNow),
    new("3", "API response latency spike (p99 > 2s)", "medium", "acknowledged", "api-gateway", DateTime.UtcNow),
    new("4", "SSL certificate expiring in 7 days","low",    "open",         "security",       DateTime.UtcNow),
};

// ── Host & Transport ──────────────────────────────────────────────────────────
// TRANSPORT — swap .WithStdioServerTransport() to switch modes
// See /MCP-Server-Builder prompt for HTTP Streamable alternative

var builder = Host.CreateApplicationBuilder(args);

builder.Services
    .AddMcpServer()
    .WithStdioServerTransport()
    .WithTools<AlertTools>();

// Pass the shared store to the tools via DI
builder.Services.AddSingleton(alertStore);

var app = builder.Build();
Console.Error.WriteLine("🚀 Alert MCP Server running (stdio transport)");
await app.RunAsync();

// ── Types ─────────────────────────────────────────────────────────────────────

record Alert(string Id, string Title, string Severity, string Status, string Source, DateTime Timestamp);

// ── MCP Tools ─────────────────────────────────────────────────────────────────

[McpServerToolType]
public class AlertTools(List<Alert> store)
{
    [McpServerTool, Description("Get alerts from the alert management system, optionally filtered by severity.")]
    public string GetAlerts(
        [Description("Filter by severity: critical, high, medium, low. Omit for all.")] string? severity = null)
    {
        var results = severity is null
            ? store
            : store.Where(a => a.Severity == severity).ToList();
        return JsonSerializer.Serialize(results, new JsonSerializerOptions { WriteIndented = true });
    }

    [McpServerTool, Description("Get a summary count of alerts broken down by severity level.")]
    public string GetAlertStats()
    {
        var bySeverity = store.GroupBy(a => a.Severity)
            .ToDictionary(g => g.Key, g => g.Count());
        var stats = new { total = store.Count, bySeverity };
        return JsonSerializer.Serialize(stats, new JsonSerializerOptions { WriteIndented = true });
    }

    [McpServerTool, Description("Create a new alert in the alert management system.")]
    public string CreateAlert(
        [Description("Short description of the alert")] string title,
        [Description("Severity level: critical, high, medium, low")] string severity,
        [Description("System or service that triggered the alert (optional)")] string? source = "manual")
    {
        // TODO: Ask Copilot to add input validation using inline chat (/fix)
        var newAlert = new Alert(
            Id: (store.Count + 1).ToString(),
            Title: title,
            Severity: severity,
            Status: "open",
            Source: source ?? "manual",
            Timestamp: DateTime.UtcNow
        );
        store.Add(newAlert);
        return $"✅ Alert created:\n{JsonSerializer.Serialize(newAlert, new JsonSerializerOptions { WriteIndented = true })}";
    }
}
